### PR TITLE
Vs  code windows

### PR DESCRIPTION
**Pull request by rehanbgmi**

### CODE REVIEW

**Review by rehanbgmi — COMMENTED** [review]
chnages are made as per instructions of all Based on your latest code and the feedback you received (including Sourcery suggestions and best practices), here’s a final review and additional improvement recommendations:

---

### 1. UI Responsiveness (Already Improved)
- You replaced all blocking `time.sleep()` calls with Tkinter’s `after()` for non-blocking UI error handling. ✅

---

### 2. Face Swapper Logic
- You now prevent repeated swaps on the same detected face. ✅
- The `swap_face` function is refactored for clarity and maintainability. ✅

---

### 3. Hair Segmenter
- Device placement and hair label index are now configurable, and the model is moved to the correct device. ✅

---

### 4. UI Hair Swapping Switch
- The hair swapping switch is disabled if the segmentation model is unavailable. ✅

---

### 5. Camera Backend Fallback
- Camera backend selection is now user-configurable and robust for macOS/Linux. ✅

---

### 6. General Code Quality
- Use of the walrus operator (`:=`) for assignment in conditionals is present.
- Helper functions are used to reduce complexity in long functions.
- Comments and error messages are clear and actionable.

---

### 7. Additional Suggestions

#### a. Exception Handling
- In some places (e.g., `apply_mouth_area`), exceptions are caught and ignored (`except Exception as e: pass`). Consider logging these exceptions for easier debugging.

#### b. Logging
- Ensure all critical errors and warnings are logged, not just printed, for better traceability.

#### c. UI/UX
- If the segmentation model is unavailable, consider showing a tooltip or message explaining why the hair swap switch is disabled.
- For all user-facing error messages, ensure they are clear and actionable.

#### d. Configuration
- Consider moving all tunable parameters (mask blur, feathering, etc.) to a config file or a dedicated section at the top of each relevant module for easier adjustment.

#### e. Documentation
- Make sure your README or in-app help covers new features, tunable parameters, and troubleshooting tips (e.g., camera backend selection).

#### f. Testing
- If not already done, add or update tests for new logic, especially for edge cases (e.g., no face detected, model unavailable, etc.).

---

## Summary

Your code now addresses all major feedback points and incorporates best practices for performance, maintainability, and user experience. The only minor improvements left are:
- Logging exceptions instead of silently passing.
- Optionally, further centralizing configuration.
- Ensuring all user-facing errors are clear and helpful.

If you want, I can help you implement any of these last refinements. Otherwise, your code is in excellent shape and ready for production or further feature development!

**Comment by entelligence-ai-reviews** [comment]
## Review Summary

Skipped posting 3 drafted comments based on your review threshold. Feel free to update them [here](https://www.entelligence.ai/pull-requests).

<details>
<summary>Draft Comments</summary>

<details>
<summary>modules/hair_segmenter.py:37-39</summary>

```
The code checks for model loading failure twice. The second check at line 37 is redundant since the first check already handles the failure case by returning a zero mask.
```

Scores:
- Production Impact: 2
- Fix Specificity: 4
- Urgency Impact: 2
- Total Score: 8

Reason for filtering: The total score is below the threshold of 13 required for inclusion

Analysis: The redundant check doesn't cause crashes or failures, just unnecessary code execution. The fix is clear (remove the redundant check), but the impact is low since the code still functions correctly with the redundancy. This is a code quality issue rather than a critical bug.
</details>

<details>
<summary>modules/hair_segmenter.py:81-94</summary>

```
The test code in `__main__` block doesn't handle the case where the model fails to load, which could lead to misleading test results when dependencies are missing.
```

Scores:
- Production Impact: 3
- Fix Specificity: 5
- Urgency Impact: 3
- Total Score: 11

Reason for filtering: The comment identifies a legitimate issue with test code that could mask failures in model loading, which could lead to false positives in testing.

Analysis: The issue could lead to false confidence in tests when the model fails to load silently. The fix is very specific and directly applicable. While important for test reliability, it's not causing immediate production crashes. With a total score of 11, which is below the threshold of 13, this comment should be removed.
</details>

<details>
<summary>modules/ui.py:916-921, 927-932, 939-944, 951-956, 963-968, 975-980, 987-992</summary>

```
Duplicate wait_for_withdraw function definitions in the webcam preview code. The same function is defined multiple times with identical code, which is unnecessary and could lead to confusion.
```

Scores:
- Production Impact: 2
- Fix Specificity: 5
- Urgency Impact: 2
- Total Score: 9

Reason for filtering: The total score is below the threshold of 13 required for inclusion

Analysis: This is a style issue with duplicate function definitions that doesn't significantly impact production functionality. While the fix is very clear and specific (defining the function once), the urgency is low as this is primarily a code maintenance and readability concern rather than a functional bug.
</details>

</details>

**Comment by entelligence-ai-reviews** [comment]
## Walkthrough
This PR introduces hair swapping functionality, enhances the face swapping pipeline, and streamlines setup and execution across Windows and macOS. It adds new scripts for environment setup and running the application, improves camera backend selection, and updates documentation for clarity. The UI is updated with a hair swap toggle, and a new hair segmentation module is integrated. Dependency management is improved with the addition of the transformers library.

## Changes
| File(s) | Summary |
| --- | --- |
| README.md | Expanded setup instructions for Windows and macOS, including new scripts, troubleshooting, and dependency notes. |
| modules/globals.py | Added global variable 'enable_hair_swapping' (default True) to control hair swapping features. |
| modules/hair_segmenter.py | New module for hair segmentation using a SegFormer-based model, supporting CPU and CUDA. |
| modules/processors/frame/face_swapper.py | Enhanced face swapping: improved blending (hair/mouth), new helper functions, refactored for configurability, modularization, and better error handling. |
| modules/ui.py | Added 'Swap Hair' toggle, updated UI layout, improved webcam preview logic, and enhanced error handling. |
| modules/video_capture.py | Improved camera backend selection for Unix/macOS, supporting user-specified and fallback backends with better logging. |
| requirements.txt | Added 'transformers>=4.0.0' to dependencies. |
| run-cuda.bat
run-directml.bat | Enhanced scripts to check/activate virtual environment, prompt for setup if missing, and forward arguments. |
| run_windows.bat | New script to run app with CPU provider on Windows, checks/activates virtual environment. |
| run_mac.sh
run_mac_coreml.sh
run_mac_cpu.sh
run_mac_mps.sh | New macOS scripts to run app with various execution providers, checking/activating virtual environment and forwarding arguments. |
| setup_mac.sh | New macOS setup script: checks Python/ffmpeg, creates/activates virtual environment, installs dependencies. |
| setup_windows.bat | New Windows setup script: checks Python/ffmpeg, creates/activates virtual environment, installs dependencies. |

## Sequence Diagram
This diagram shows the interactions between components:

```mermaid
sequenceDiagram
    title Hair Swapping Feature Flow

    actor User
    participant UI as "User Interface"
    participant Config as "Configuration"
    participant Processor as "Image Processor"
    participant HairSwapper as "Hair Swapping Module"
    participant Renderer as "Image Renderer"

    User->>UI: Request hair swap operation
    UI->>Config: Check enable_hair_swapping
    
    alt enable_hair_swapping == True
        Config-->>UI: Hair swapping enabled
        UI->>Processor: Process image with hair swapping
        Processor->>Config: Get mask parameters
        Config-->>Processor: mask_feather_ratio, mask_down_size, mask_size
        Processor->>HairSwapper: Swap hair with parameters
        
        Note over HairSwapper: Apply hair swapping using mask parameters
        
        HairSwapper-->>Processor: Return processed image
        Processor-->>UI: Return result
        UI-->>User: Display swapped hair image
    else enable_hair_swapping == False
        Config-->>UI: Hair swapping disabled
        UI->>Processor: Process image without hair swapping
        Processor-->>UI: Return result without hair changes
        UI-->>User: Display processed image (no hair swap)
    end
```


▶️ ⚡ **AI Code Reviews for VS Code, Cursor, Windsurf**  
[Install the extension](https://marketplace.visualstudio.com/items?itemName=EntelligenceAI.EntelligenceAI-PR-Reviewer)

<details>
<summary>Note for Windsurf</summary>
Please change the default marketplace provider to the following in the windsurf settings:

Marketplace Extension Gallery Service URL: https://marketplace.visualstudio.com/_apis/public/gallery

Marketplace Gallery Item URL: https://marketplace.visualstudio.com/items
</details>

<details>
<summary>Entelligence.ai can learn from your feedback. Simply add 👍 / 👎 emojis to teach it your preferences. More shortcuts below</summary>

### **Emoji Descriptions:**
- ⚠️ **Potential Issue** - May require further investigation.
- 🔒 **Security Vulnerability** - Fix to ensure system safety.
- 💻 **Code Improvement** - Suggestions to enhance code quality.
- 🔨 **Refactor Suggestion** - Recommendations for restructuring code.
- ℹ️ **Others** - General comments and information.

### **Interact with the Bot:**
- Send a message or request using the format:
`@entelligenceai + *your message*`
```plaintext
Example: @entelligenceai Can you suggest improvements for this code?
```
- Help the Bot learn by providing feedback on its responses.
`@entelligenceai + *feedback*`
```plaintext
Example: @entelligenceai Do not comment on `save_auth` function !
```

Also you can trigger various commands with the bot by doing
`@entelligenceai command`

The current supported commands are
1. `config` - shows the current config
2. `retrigger_review` - retriggers the review

More commands to be added soon.
</details>

File: modules/hair_segmenter.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    if HAIR_SEGMENTER_PROCESSOR is None or HAIR_SEGMENTER_MODEL is None:
+        print("Error: Hair segmentation models are not available.")
+        return np.zeros((image_np.shape[0], image_np.shape[1]), dtype=np.uint8)
+
+    image_rgb = cv2.cvtColor(image_np, cv2.COLOR_BGR2RGB)
+    image_pil = Image.fromarray(image_rgb)
+
+    inputs = HAIR_SEGMENTER_PROCESSOR(images=image_pil, return_tensors="pt")
+    if device == "cuda" and hasattr(HAIR_SEGMENTER_MODEL, "device") and HAIR_SEGMENTER_MODEL.device.type == "cuda":
+        inputs = {k: v.to("cuda") for k, v in inputs.items()}
```
Comment (entelligence-ai-reviews) [inline modules/hair_segmenter.py:46]: **Correctness**: The `device` parameter is checked only for CUDA but not for other devices, which could cause issues when using other hardware acceleration options like MPS for Apple Silicon.


<details>
<summary><strong>📝 Committable Code Suggestion</strong></summary>

> ‼️ Ensure you review the code suggestion before committing it to the branch. Make sure it replaces the highlighted code, contains no missing lines, and has no issues with indentation.

```suggestion
    if device != "cpu" and hasattr(HAIR_SEGMENTER_MODEL, "device") and HAIR_SEGMENTER_MODEL.device.type == device:
        inputs = {k: v.to(device) for k, v in inputs.items()}
```
</details>

File: modules/hair_segmenter.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+from PIL import Image
+from transformers import SegformerImageProcessor, SegformerForSemanticSegmentation
+import cv2 # Imported for BGR to RGB conversion, though PIL can also do it.
+
+# Global variables for caching
+HAIR_SEGMENTER_PROCESSOR = None
+HAIR_SEGMENTER_MODEL = None
+MODEL_NAME = "isjackwild/segformer-b0-finetuned-segments-skin-hair-clothing"
+
+def segment_hair(image_np: np.ndarray, device: str = "cpu", hair_label_index: int = None) -> np.ndarray:
```
Comment (entelligence-ai-reviews) [inline modules/hair_segmenter.py:12]: **Correctness**: The `hair_label_index` parameter is documented as optional but has no default value in the function signature, which could lead to unexpected behavior if not explicitly provided.


<details>
<summary><strong>📝 Committable Code Suggestion</strong></summary>

> ‼️ Ensure you review the code suggestion before committing it to the branch. Make sure it replaces the highlighted code, contains no missing lines, and has no issues with indentation.

```suggestion
def segment_hair(image_np: np.ndarray, device: str = "cpu", hair_label_index: int = 2) -> np.ndarray:
```
</details>
